u_collate.c

```diff
@@ -80,8 +114,50 @@ int u_strcasecoll0(const char *str1, const char *str2)
 	return u_strcasecoll(str1, str2);
 }
 
+/* Helper function to create collation key using CoreFoundation on macOS */
+#ifdef __APPLE__
+static char *cf_create_collation_key(const char *str) 
+{
+	char *result = NULL;
+	CFStringRef cfStr = CFStringCreateWithCString(NULL, str, kCFStringEncodingUTF8);
+	
+	if (cfStr) {
+		/* Get a representation that can be used for sorting */
+		CFMutableStringRef mStr = CFStringCreateMutableCopy(NULL, 0, cfStr);
+		if (mStr) {
+			/* Perform canonical decomposition and strip combining marks for comparison */
+			CFStringNormalize(mStr, kCFStringNormalizationFormD);
+			
+			/* Convert back to C string */
+			size_t max_size = CFStringGetMaximumSizeForEncoding(CFStringGetLength(mStr), kCFStringEncodingUTF8) + 2;
+			result = xmalloc(max_size);
+			result[0] = 'A'; /* Mark as preprocessed */
```

Comment:
Does this `result[0]` addition actually change anything?

The existing `u_strcoll_key()` seems to do this to differentiate the output between the two cases (the top `using_utf8` path which doesn't prepend the output with an `A` and the second path which does).  This seems to just exist so that there will be some sorting difference between things which were already UTF-8 and ones that required some normalisation first.

The point of the `_key()` functions are to just produces a representative string that can be used with `strcmp()` to return how the original input strings should compare in the active locale.  Therefore if every string that it produces has an `A` prepended should act the same as if none of them did, no?